experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-translator.ts

```diff
@@ -1,11 +1,185 @@
+ * 4. Makes the EIO `sid` same as its corresponding Azure Web PubSub client connection id.
  */
-class WebPubSubTranslator {
-}
+export class WebPubSubTranslator {
+  /**
+   * Each `WebPubSubTranslator` instance is bound to a Engine.IO server instance and vice versa.
+   */
+  public linkedEioServer: BaseServer;
+
+  /**
+   * Client for connecting to a Web PubSub hub
+   */
+  private serviceClient: WebPubSubServiceClient;
+
+  /**
+   * Map from the `connectionId` of each client to its corresponding logical `ClientConnectionContext`.
+   */
+  private clientConnections: Map<string, ClientConnectionContext> = new Map();
+
+  /**
+   * Handle upstream invoke requests from AWPS.
+   */
+  private webPubSubEventHandler: WebPubSubEventHandler;
+
+  /**
+   * Options for Azure Web PubSub service.
+   */
+  private webPubSubOptions: WebPubSubExtensionOptions;
+
+  /**
+   * In native Engine.IO, the `sid` of each EIO connnection is generated by server randomly.
+   * As for AWPS, it generates `ConnectionId` for each client.
+   * For each EIO connection, the extension enforces its `sid` in server side is same as the ConnectionId assigned by service.
+   * This array stores all `ConnectionId` which is generated by AWPS and is prepared to be assigned to EIO connection.
+   */
+  private candidateSids: Array<string> = new Array();
+
+  constructor(server: BaseServer, options: WebPubSubExtensionOptions) {
+    debug("constructor");
+
+    this.linkedEioServer = server;
+    this.webPubSubOptions = options;
+    this.serviceClient = new WebPubSubServiceClient(
+      this.webPubSubOptions.connectionString,
+      this.webPubSubOptions.hub,
+      {
+        allowInsecureConnection: true,
+      }
+    );
+
+    this.webPubSubEventHandler = new WebPubSubEventHandler(this.webPubSubOptions.hub, {
+      path: this.webPubSubOptions.path,
+      handleConnect: async (req, res) => {
+        var connectionId = req.context.connectionId;
+        debug(`onConnect, connectionId = ${connectionId}`);
+
+        var context = new ClientConnectionContext(this.serviceClient, connectionId);
+
+        context.on(EIO_SERVER_ACCEPT_CONNECTION, (openPacketPayload: string) => {
+          let payloadWithoutType = openPacketPayload.substring(1);
+          res.success({
+            socketioHandshake: JSON.parse(payloadWithoutType),
+          } as WebPubSubConnectResponse);
+        });
+
+        context.on(EIO_SERVER_REFUSE_CONNECTION, (errorMessage: string) => {
+          res.fail(400, `EIO server refused connection with error: ${errorMessage}`);
+        });
 
-export { WebPubSubTranslator };
+        var connectReq = this.convertWebPubsubConnectReqToEioHandshakeReq(req, context);
+
+        this.candidateSids.push(connectionId);
+        this.clientConnections.set(connectionId, context);
+        // @ts-ignore to access private `handshake` method
+        await this.linkedEioServer.handshake(WEBPUBSUB_TRANSPORT_NAME, connectReq);
```

Comment:
can it be guaranteed that the response is always set? Otherwise, it results in the response hung. Also after res.fail/success will there be other code executions inside this `handshake`? what if that code execution fails?

Reply:
Add a `try-catch` and a timeout checker to avoid response hung here